File: src/actfuncs.c
#include <stdio.h> // BUFSIZ
#include <string.h> // strchr()
#include <assert.h> // assert()
#include <math.h> // floor()
#include "common.h" // Map
#include "errorexit.h" // die()
#include "helperfunctions.h" // jalloc()

#define EPS 0.0000001

static Map life_tables;

enum {MAX_AGE = 512};

struct life_table {
	size_t size;
	unsigned lx[MAX_AGE];
};

static const struct life_table *make_life_table(const char *name)
{ 
	const char *lp = 0;
	char line[BUFSIZ];
	FILE *lt = 0;
	size_t age = 0;

	if (0 == (lt = fopen(name, "r"))) {
		die("can't open %s", name);
	}

	struct life_table *life_table = jalloc(1, sizeof(*life_table));

	while((fgets(line, BUFSIZ, lt))) {
		if (!(lp = strchr(line, ','))) {
			die("file '%s' has incorrect format", name);
		}
		if (age < MAX_AGE) {
			life_table->lx[age++] = atoi(++lp);
		}
	}

	fclose(lt);

	life_table->size = age;
	map_put(&life_tables, name, life_table);
	return life_table;
}


unsigned lx(const char *name, size_t age)
{
	assert(name);
	const struct life_table *life_table = map_get(&life_tables, name);
	if (!life_table) {
		life_table = make_life_table(name);
		assert(life_table);
		printf("life table '%s' created\n", name);
	}

	if (age > life_table->size - 1) {
		return 0;		
	} else {
		return life_table->lx[age];
	}
}

double npx(const char *table, double ageX, double ageXn, int corr)
{
	ageX += corr;
	ageXn += corr;
	if (ageX == ageXn) {
		return 1.0;
	} else if (ageX > ageXn) {
		return 0.0;
	} else if (ageX < 0 || ageXn < 0) {
		return 0.0;
	}

	double ip1 = 0.0;
	double ip2 = 0.0;
	unsigned lxX = 0;
	unsigned lxX1 = 0;
	unsigned lxXn = 0;
	unsigned lxXn1 = 0;
	lxX = lx(table, (size_t)ageX);
	lxX1 = lx(table, (size_t)(ageX + 1));
	lxXn = lx(table, (size_t)ageXn);
	lxXn1 = lx(table, (size_t)(ageXn + 1));

	ip1 = lxX - (ageX - floor(ageX)) * (lxX - lxX1);
	ip2 = lxXn - (ageXn - floor(ageXn)) * (lxXn - lxXn1);

	if (0 == ip1) {
		assert(0 == ip2);
		return 0.0;
	} else {
		return ip2/ip1;
	}
}

double nEx(const char *table, double i, double charge, double ageX,
		double ageXn, int corr)
{
	double im = (1 + i)/(1 + charge) - 1;
	double n = ageXn - ageX;
	double vn = 1 / pow(1 + im, n);
	double nPx = npx(table, ageX, ageXn, corr);
	return vn * nPx;
}

/*
 * annuity immediate (prepost == 1) / due (prepost == 0). if 12 is not
 * divisible by 12 then return -1.0 to indicate incorrect input.
 */
double axn(const char *table, double i, double charge, unsigned prepost,
		unsigned term, double ageX, double ageXn, int corr)
{
	int payments = 0;
	double ageXk = 0.0;
	double value = 0.0;
	double termfrac = 0.0;

	if (ageX > ageXn + EPS) {
		value = 0;
	} else if (12 % term) {
		return -1.0;
	} else {
		termfrac = 1.0 / term;
		ageXk = ageX + (double)prepost/term;
		payments = (ageXn - ageX) * term + EPS;
		while (payments--) {
			value += nEx(table, i, charge, ageX, ageXk, corr);
			ageXk += termfrac;
		}
	}

	return value;
}

/* 
 * nAx = v^(1/2)*1Qx + v^(1+1/2)*1Px*1q_{x+1} + ...
 * + v^(n-1+1/2)*{n-1}_Px*1Q_{x+n-1}
 */
double Ax1n(const char *table, double i, double charge, double ageX,
		double ageXn, int corr)
{
	int k = 0;
	int payments = 0;
	double im = 0.0;
	double v = 0.0;
	double value = 0.0;
	double kPx = 0.0;
	double qx = 0.0;

	if (ageX > ageXn + EPS) {
		return 0;
	} else {
		im = (1 + i)/(1 + charge) - 1;
		v = 1/(1 + im);
		payments = ageXn - ageX + EPS;

		while (payments--) {
			kPx = npx(table, ageX, ageX + k, corr);
			qx = (1 - npx(table, ageX + k, ageX + k + 1, corr));
			value += pow(v, k + 1.0/2) * kPx * qx;
			k++;
		}
		kPx = npx(table, ageX, ageX + k, corr);
		qx = (1 - npx(table, ageX + k, ageXn, corr));
		value += pow(v, k + 1.0/2) * kPx * qx;

		return value;
	}
}

// IAx1n = sum^{n-1}_k=1:k*1A_{x+k}*kEx
double IAx1n(const char *table, double i, double charge, double ageX,
		double ageXn, int corr)
{
	int k = 1;
	int payments = 0;
	double value = 0.0;

	if (ageX > ageXn + EPS) {
		return 0;
	} else {
		payments = ageXn - ageX + EPS;
		while (payments--) {
			value += k * Ax1n(table, i, charge, ageX + k - 1,
					ageX + k, corr)
				* nEx(table, i, charge, ageX, ageX + k - 1, corr);
			k++;
		}
		value += k * Ax1n(table, i, charge, ageX + k - 1,
				ageXn, corr)
			* nEx(table, i, charge, ageX, ageXn, corr);

		return value;
	}
}

/* 
 * This function hasn't been completed because I don't think I need it.
 * At the moment it only works for term = 1
 */
double Iaxn(const char *table, double i, double charge, unsigned prepost,
		unsigned term, double ageX, double ageXn, int corr)
{
	int k = 1;
	int payments = 0;
	double ageXk = 0.0;
	double value = 0.0;

	if (ageX > ageXn + EPS) {
		return 0;
	} else {
		ageXk = ageX + (double)prepost/term;
		payments = (ageXn - ageX) * term + EPS;
		while (payments--) {
			value += k++ * nEx(table, i, charge, ageX, ageXk, corr);
			ageXk += 1.0/term;
		}
		ageXk -= 1.0/term * prepost;
		value /= term;
		value += (ageXn - ageXk) * k
			* nEx(table, i, charge, ageX,
					((int)(ageXn*term + EPS))/term
					+ term*prepost, corr);

		return value;
	}
}

File: python/actfuncs.py
path = "/home/doctormartin67/Coding/tables/tables/"

EPS = 0.0000000001

def calc_years(d1, d2, extra_month):
	return d1.year - d2.year + (d1.month - d2.month - extra_month)/12

def create_table(table):
	f = open(path + table, "r")
	new_table = [int(s.split(",")[1].rstrip()) for s in f]
	return new_table

def lx(table, age):
	if not table in lx.tables:
		lx.tables[table] = create_table(table)

	if age > len(lx.tables[table]) - 1:
		return 0
	elif age < 0:
		return 0
	else:
		return lx.tables[table][age]
lx.tables = {}

def npx(table, ageX, ageXn, corr):
	ageX += corr
	ageXn += corr
	if ageX >= ageXn:
		return 1.0
	elif ageX < 0 or ageXn < 0:
		return 0.0

	lxX = lx(table, int(ageX))
	lxX1 = lx(table, int(ageX + 1))
	lxXn = lx(table, int(ageXn))
	lxXn1 = lx(table, int(ageXn + 1))

	ip1 = lxX - (ageX - int(ageX)) * (lxX - lxX1)
	ip2 = lxXn - (ageXn - int(ageXn)) * (lxXn - lxXn1)

	if 0 == ip1:
		assert(0 == ip2)
		return 0.0
	else:
		return ip2/ip1

def nEx(table, i, charge, ageX, ageXn, corr):
	if (ageX == ageXn):
		return 1
	im = (1 + i)/(1 + charge) - 1
	n = ageXn - ageX
	vn = 1 / (1 + im) ** n
	nPx = npx(table, ageX, ageXn, corr)
	return vn * nPx
	
def axn(table, i, charge, prepost, term, ageX, ageXn, corr):
	if ageX > ageXn + EPS:
		return 0
	elif 12 % term:
		return -1.0
	else:
		value = 0.0
		termfrac = 1.0 / term
		ageXk = ageX + prepost/term
		payments = int((ageXn - ageX) * term + EPS)
		while (payments):
			value += nEx(table, i, charge, ageX, ageXk, corr)
			ageXk += termfrac
			payments -= 1

	return value

def Ax1n(table, i, charge, ageX, ageXn, corr):
	if ageX > ageXn + EPS:
		return 0
	else:
		k = 0
		value = 0.0
		im = (1 + i)/(1 + charge) - 1
		v = 1/(1 + im)
		payments = int(ageXn - ageX + EPS)

		while payments:
			kPx = npx(table, ageX, ageX + k, corr)
			qx = (1 - npx(table, ageX + k, ageX + k + 1, corr))
			value += v ** (k + 1.0/2) * kPx * qx
			k += 1
			payments -= 1
		kPx = npx(table, ageX, ageX + k, corr)
		qx = (1 - npx(table, ageX + k, ageXn, corr))
		value += v (k + 1.0/2) * kPx * qx

		return value

def IAx1n(table, i, charge, ageX, ageXn, corr):
	if ageX > ageXn + EPS:
		return 0
	else:
		k = 1
		value = 0.0
		payments = int(ageXn - ageX + EPS)
		while payments:
			Ax1 = Ax1n(table, i, charge, ageX + k-1, ageX+k, corr)
			Ex = nEx(table, i, charge, ageX, ageX + k - 1, corr)
			value += k * Ax1 * Ex
			k += 1
			payments -= 1

		Ax1 = Ax1n(table, i, charge, ageX + k - 1, ageXn, corr)
		value += k * Ax1 * nEx(table, i, charge, ageX, ageXn, corr)

		return value

# This function hasn't been completed because I don't think I need it.
# At the moment it only works for term = 1
def Iaxn(table, i, charge, prepost, term, ageX, ageXn, corr):
	if ageX > ageXn + EPS:
		return 0
	else:
		k = 1
		value = 0.0
		ageXk = ageX + prepost/term
		payments = int((ageXn - ageX) * term + EPS)
		while payments: 
			value += k * nEx(table, i, charge, ageX, ageXk, corr)
			ageXk += 1.0/term
			k += 1
			payments -= 1

		return value
	
def lump_sum(combination, res, prem, pre_post, term, age, NRA, table,
	i, charge, admin_cost, delta_cap, dth_cap, ps_is_mixed, X10, b2):
	if age == NRA:
		return 0.0
	
	if 0 == res and 0 == prem:
		return 0.0

	ax = axn(table, i, charge, pre_post, term, age, NRA, 0)
	Ex = nEx(table, i, charge, age, NRA, 0)

	if "UKMS" == combination or "UKZT" == combination:
		lump_sum = (res + prem * (1 - admin_cost) * ax) / Ex
	elif "MIXED" == combination:
		Ax1 = Ax1n(table, i, charge, age, NRA, 0)
		ax_cost = axn(table, i, charge, 0, 1, age, NRA, 0)
		lump_sum = (res + prem * (1 - admin_cost) * ax)
		if ps_is_mixed:
			lump_sum /= (Ex + 1/X10 * Ax1 + 1/X10 * b2 * ax_cost)
		else:
			lump_sum /= Ex
	elif "UKMT" == combination:
		Ax1 = Ax1n(table, i, charge, age, NRA, 0)
		ax_cost = axn(table, i, charge, 0, 1, age, NRA, 0)
		IAx1 = IAx1n(table, i, charge, age, NRA, 0)
		Iax = Iaxn(table, i, charge, 0, 1, age, NRA, 0)
		prem *= (1 - admin_cost)
		lump_sum = (res + prem * ax - dth_cap * (Ax1 + b2 * ax_cost) 
			- prem * (IAx1 + b2 * Iax)) / Ex
	else:
		assert(0)

	return lump_sum + delta_cap * (NRA - age) * 12

def reserves(combination, cap, prem, pre_post, term, age, NRA, table,
	i, charge, admin_cost, delta_cap, dth_cap, ps_is_mixed, X10, b2):
	if age == NRA:
		return 0.0
	
	if 0 == cap and 0 == prem:
		return 0.0

	ax = axn(table, i, charge, pre_post, term, age, NRA, 0)
	Ex = nEx(table, i, charge, age, NRA, 0)

	if "UKMS" == combination or "UKZT" == combination:
		return (cap - delta_cap * (NRA - age) * 12) * Ex \
			- prem * (1 - admin_cost) * ax
	elif "MIXED" == combination:
		Ax1 = Ax1n(table, i, charge, age, NRA, 0)
		ax_cost = axn(table, i, charge, 0, 1, age, NRA, 0)
		if ps_is_mixed:
			reserves = cap * (Ex + 1/X10 * Ax1
				+ 1/X10 * b2 * ax_cost) 
		else:
			reserves = cap * Ex
		return reserves - prem * (1 - admin_cost) * ax
	elif "UKMT" == combination:
		Ax1 = Ax1n(table, i, charge, age, NRA, 0)
		ax_cost = axn(table, i, charge, 0, 1, age, NRA, 0)
		IAx1 = IAx1n(table, i, charge, age, NRA, 0)
		Iax = Iaxn(table, i, charge, 0, 1, age, NRA, 0)
		prem *= (1 - admin_cost)
		return cap * Ex - prem * ax + dth_cap * (Ax1 + b2 * ax_cost) \
			+ prem * (IAx1 + b2 * Iax)
	else:
		assert(0)


##############################
# below are all the unit tests
##############################

tables = ["LXMR", "LXFR", "LXMK", "LXFK'", "Lxnihil"]
def test_lx():
	assert(lx("LXFR", 40) == 982954)
	assert(lx("LXFR", 0) == 1000000)
	assert(lx("LXFR", 120) == 0)
	assert(lx("Lxnihil", 0) == 1000000)
	assert(lx("Lxnihil", 1) == 1000000)

def test_npx(i):
	for table in tables:
		nPx = npx(table, i%110, (i+1)%110, 0)
		nPx = npx(table, i%107, (i+1)%107, 3)
		nPx = npx(table, i%110, (i+1)%110, -3)
		assert(1 == npx(table, 40, 40, 0))
		assert(0 == npx(table, 0, 130, 0))
		age1 = i%110 - 3
		age2 = (i+1)%110 - 3
		lx1 = lx(table, age1)
		lx2 = lx(table, age2)
		if age2 < age1:
			continue
		if 0 == lx1:
			assert(0 == nPx)
		else:
			assert(abs(nPx - lx2 / lx1) < EPS)

def test_nEx(i):
	for table in tables:
		ageX = i%110
		ageXn = (i+1)%110
		nex = nEx(table, 0.0, 0.0, ageX, ageXn, -3)
		assert(abs(nex - npx(table, ageX, ageXn, -3)) < EPS)
		nex = nEx(table, 0.01, 0.0, ageX, ageXn, -3)
		assert(abs(nex - npx(table, ageX, ageXn, -3)
					* (1 + 0.01) ** (ageX - ageXn)) < EPS)
		nex = nEx(table, 0.012, 0.001, ageX, ageXn, -3)
		assert(abs(nex - npx(table, ageX, ageXn, -3)
					* ((1 + 0.012) / (1 + 0.001))
						** (ageX - ageXn)) < EPS)
	
def test_axn(i):
	for table in tables:
		ax = axn(table, 0.02, 0.001, 1, 12, i, i, -3)
		assert(0 == ax)
		ax = axn(table, 0.02, 0.001, 0, 1, i, i+1, -3)
		assert(1 == ax)
		ax = axn(table, 0.02, 0.001, 1, 1, i, i+1, -3)
		nex = nEx(table, 0.02, 0.001, i, i+1, -3)
		assert(abs(nex - ax) < EPS)

def test_all():
	N = 1024 
	for i in range(0, N):
		test_lx()
		test_npx(i)
		test_nEx(i)
		test_axn(i)

if __name__ == "__main__":
	test_all()
